bufit.go

```diff
@@ -131,6 +131,9 @@ func (b *Buffer) NextReader() io.ReadCloser {
 // Write appends the given data to the buffer. All active readers will
 // see this write.
 func (b *Buffer) Write(p []byte) (int, error) {
+	if !b.alive() {
+		return 0, io.ErrClosedPipe
+	}
```

Comment:
No harm really, though we can actually keep the b.alive() check outside the critical section as well. This lets us do a quick check without locking (shortcut to failure case if its already dead), if that check succeeds, we move into the critical section and check again to make sure its still alive (which you already added).

Also, we'll need to add a lock on the Close() section otherwise the critical section here can still happen concurrently with close.